compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -774,12 +774,7 @@ where
                         TyMaybeWithLayout::Ty(tcx.mk_mut_ptr(tcx.types.unit))
                     } else if i == 1 {
                         // FIXME(dyn-star) same FIXME as above applies here too
-                        TyMaybeWithLayout::Ty(
-                            tcx.mk_imm_ref(
-                                tcx.lifetimes.re_static,
-                                tcx.mk_array(tcx.types.usize, 3),
-                            ),
-                        )
+                        TyMaybeWithLayout::Ty(tcx.mk_imm_ptr(tcx.mk_array(tcx.types.usize, 3)))
```

Comment:
This is necessary because if we represent the vtable as a reference, then CTFE wants to validate the pointer. However, the vtable is represented by `GlobalAlloc::VTable`, which is treated like `Size::ZERO`. So it complains that we're trying to write 24 bytes into an allocation of size 0.

Alternatively, we could fix this instead I guess:

https://github.com/rust-lang/rust/blob/d962ea57899d64dc8a57040142c6b498a57c8064/compiler/rustc_const_eval/src/interpret/memory.rs#L711

Reply:
This shouldn't happen, validity checking should stop at the `dyn*` and not descend into its components.

Did you rebase? https://github.com/rust-lang/rust/pull/107728 fixed some issues with dyn* handling in the validity check.

Reply:
Yes, this is based off of a rebased master.

Reply:
Strange. Then CTFE should never see this type.

What's the error you are getting otherwise, and for which code? Ideally with RUSTC_CTFE_BACKTRACE=1

Reply:
> Strange. Then CTFE should never see this type.

Why? 

https://github.com/rust-lang/rust/pull/104694/files#diff-60d85c7ca5a99b206384a4ccb596b7e0f9824fd6606ae7241e2d4a4e5ae65025R134

Here we write the vtable to the second part of the scalar pair. So we're trying to write the vtable allocation pointer to a place of type `&'static [usize; 3]`, which then tries to validate that vtable pointer.

Reply:
Oh, that's where this comes from. But during CTFE we don't do validation on writes... this should only affect Miri. Still a problem though.

However why wouldn't this also be a problem for vtables in wide pointers?

I'm also worried using a raw pointer type here could have bad side-effects, like losing `nonnull` attributes. IIRC there are *two* places where we define the layout/type/attributes for fields of wide ptrs and `dyn*` and I have no idea what happens when they start to diverge.